src/lib/pgp-key.cpp

```diff
@@ -1023,6 +1043,94 @@ pgp_subsig_from_signature(pgp_subsig_t *subsig, const pgp_signature_t *sig)
     return false;
 }
 
+bool
+pgp_key_has_signature(const pgp_key_t *key, const pgp_signature_t *sig)
+{
+    for (size_t i = 0; i < pgp_key_get_subsig_count(key); i++) {
+        pgp_subsig_t *subsig = pgp_key_get_subsig(key, i);
+        if (signature_pkt_equal(&subsig->sig, sig)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static bool
+pgp_key_replace_rawpacket(pgp_key_t *key, pgp_rawpacket_t *oldpkt, pgp_rawpacket_t *newpkt)
+{
+    for (size_t i = 0; i < pgp_key_get_rawpacket_count(key); i++) {
+        pgp_rawpacket_t *curpkt = pgp_key_get_rawpacket(key, i);
+        if ((curpkt->length != oldpkt->length) || (curpkt->tag != oldpkt->tag)) {
+            continue;
+        }
+        if (memcmp(curpkt->raw, oldpkt->raw, curpkt->length)) {
+            continue;
+        }
+        pgp_rawpacket_t pktcp = {};
+        pktcp = *newpkt;
+        pktcp.raw = (uint8_t *) malloc(newpkt->length);
+        if (!pktcp.raw) {
+            RNP_LOG("allocation failed");
+            return false;
+        }
+        memcpy(pktcp.raw, newpkt->raw, newpkt->length);
+        pgp_rawpacket_free(curpkt);
+        *curpkt = pktcp;
+        return true;
+    }
+    return false;
+}
+
+pgp_subsig_t *
+pgp_key_replace_signature(pgp_key_t *key, pgp_signature_t *oldsig, pgp_signature_t *newsig)
+{
+    pgp_subsig_t *subsig = NULL;
+    for (size_t i = 0; i < pgp_key_get_subsig_count(key); i++) {
+        subsig = pgp_key_get_subsig(key, i);
+        if (signature_pkt_equal(&subsig->sig, oldsig)) {
+            break;
+        }
+    }
+    if (!subsig) {
```

Comment:
Non-null `subsig` here doesn't mean that `signature_pkt_equal` matched. It may be a leftover from the last iteration.

Reply:
@rrrooommmaaa Thanks for finding this out, will fix it.

Reply:
@rrrooommmaaa Should be good now.